### ISSUE TITLE
failed to find a valid digest in the 'integrity' attribute?

### ISSUE THREAD

**Issue by georgesaliba** [Baroshem/nuxt-security#297]
I recently updated to rc4 and all users got the following error when I deployed to production:

![image](https://github.com/Baroshem/nuxt-security/assets/1479319/3214950e-ac7f-4d13-a576-dd3e3dfed3e5)

Is this related to the update? Is it a bug, or would there be a changelog somewhere that would indicate what changes I might need to make?

**Comment by Baroshem** [comment]
Hello,

Yes, I think it is caused by the latest changes in RC.4 where we have enabled strict CSP by default to enforce good security practices.

Take a look at following resources to learn more about Integrity

- https://nuxt-security.vercel.app/documentation/utils/subresource-integrity#usage
- https://nuxt-security.vercel.app/documentation/advanced/strict-csp/#ssg-mode

Also, I think

**Comment by georgesaliba** [comment]
Makes sense. Yes, maybe a migration guide would be helpful - though maybe it makes sense that I should expect it to have breaking changes since I was on the RC. :)

But maybe the bigger problem is that I'm not sure that the feature is working right? That first article says "Nuxt Security automatically computes the integrity hash of each static asset (scripts, stylesheets, etc.) that are bundled in your Nuxt Application, and then inserts this value in the resulting HTML file." I'm pretty sure the failing files were all CSS & JS that are generated by my Nuxt application. Here's the screenshot with less obfuscated: 

![image](https://github.com/Baroshem/nuxt-security/assets/1479319/01de2f55-55e5-410d-bf70-a5d6127bda7b)

**Comment by Baroshem** [comment]
We are trying our best not to include any breaking changes but sometimes it is just not possible. There we also try to add migration guides to make migration easier.

At this point, if after upgrading, you encounter an issue that is blocking the applciation you can either stay at 1.0.0-rc.3 version or disable Subresource Integrity for now (security.sri: false)

Me and @vejja will try to find the solution in the meantime :)

**Comment by vejja** [comment]
@georgesaliba Can you tell me what is the value of the integrity attribute of your /_nuxt/entry.9660c2bb.js ?

**Comment by georgesaliba** [comment]
@vejja How can I find that? I think that file is dynamically constructed by Nuxt, the issue only occurred when I tried to ship the code into production (didn't repro locally), and I had to quickly roll it back because it caused the whole site to go down. So even if I knew how to calculate it, would I have to cause another live site outage to find that value, or is there a less impactful way?

**Comment by georgesaliba** [comment]
Yes, I can run build locally.

But not sure how to find the integrity. I tried the fourth option command from Baroshem's second link above on my generated file, and it gave me a hash of wsVcd+MufiHE2rN83nQXLvySQzpPqUfo/pALNeuhqmHpP/M9kO0/4Yw8IVkyvW7g - but I'm not sure what the production one is (the code has changed since the outage). I'm guessing I'd have to deploy again to find it? Also assuming if I were to deploy with sri: false, I could then just download the file from my production server and run the same command to see if it's the same? There's a big launch today, so I probably won't be able to experiment with it this weekend, but maybe I can try this in the middle of the night sometime next week.

To answer your question, my stack is all based in Google Cloud, fronted by Cloudflare. So the static files are hosted by Google Firebase and dynamic bits are served by Google Cloud Functions, and Cloudflare caches various things for different periods. I'm not aware of any sort of minification or such happening post build, but I guess I wouldn't be shocked if either Google or Cloudflare was trying to optimize in some way, adding additional headers, etc.

**Comment by vejja** [comment]
Ok thank you very much for the explanation.

You will find the integrity hash in the <head> section of the html of your file. Open your website with Chrome DevTools, and look either in the Elements tab (find the <head> in the parsed document), or in the Network tab (look for the Response sub-tab to find out the actual HTML that was delivered). In both cases your `/_nuxt/entry.9660c2bb.js` link in the <head> tab will have an `integrity` attribute - this is what I'm looking for.

One thing that you can do without risking another redeploy, is to save this `/_nuxt/entry.9660c2bb.js` file that Google Cloud/Cloudflare delivers to your users, and then compare it with the one that is produced by your local build.

I would want to understand if they are similar or if there are any differences. You can do a simple diff in the terminal to find out.

Totally understand that you probably don't have so much time to do that this kind of research, but if you manage to get the information it would be extremely helpful.

**Comment by georgesaliba** [comment]
I assume there's no way to get the matching values on a local machine, eh? This is the local HTML I see, but it doesn't have this integrity value:
![image](https://github.com/Baroshem/nuxt-security/assets/1479319/286640e2-d773-4ab4-bdab-ed6662273ff5)

But it also doesn't include the URL hash, so I'm guessing it's just a dev-only version, and wouldn't match my prod env entirely. Once I'm able to redeploy to prod I'll try to follow the steps you mentioned to see if I can get the right value.

Oh, I also didn't answer you earlier question - the site is a mix of SSR and SPA.

**Comment by vejja** [comment]
Yes this is a dev build, but weird that you don’t have integrity there
Locally you could try to set
- rc4 for Nuxt-security
- sri: true
- npm run generate (assuming this is your build command)
- npx serve .output/… (whatever the previous command says you should do)
And look for the html on localhost

**Comment by Baroshem** [comment]
@georgesaliba could you verify if the issue still occurs?

I would like to release a new version this week and if this issue appears for more use cases, then we have to fix it before releasing the stable 1.0.0 version

**Comment by georgesaliba** [comment]
So I was able to deploy to production with sri: false and it didn't bring the site down. However, I also don't see the integrity in the production HTML (maybe as a result of the sri being false?).

**Comment by Baroshem** [comment]
So if I understand correctly, when sri is set to true (which is by default) your site does not work correctly?

**Comment by georgesaliba** [comment]
Yes, exactly. None of the generated JS or CSS files load.

**Comment by Baroshem** [comment]
I personally cannot reproduce it unfortunately and when I was testing it before releasing and after the release the default SRI was not causing any problems for me.

@georgesaliba Could you provide us with more info about your config? TBH, I have no idea what could be causing it so any information you could give us would help us solve the issue.

Also, have you tried this from @vejja ? https://github.com/Baroshem/nuxt-security/issues/297#issuecomment-1816702939

**Comment by vejja** [comment]
> So I was able to deploy to production with sri: false and it didn't bring the site down. However, I also don't see the integrity in the production HTML (maybe as a result of the sri being false?).

Right, `sri: false` suppresses the integrity from the HTML, so this is normal.

One thing you could do, now that you have redeployed, is to `diff` the online version of your `/_nuxt/entry.xxxxxxxx.js` file with the one that you obtain locally by running `npm run generate` ?

**Comment by georgesaliba** [comment]
So I can see there are differences, and it looks like something is slightly minifying the production file, removing unneeded whitespace:

<img width="1188" alt="image" src="https://github.com/Baroshem/nuxt-security/assets/1479319/4a86494e-e5b6-4fed-850f-2d6ee807ac76">

Or this extra line feed at the end of the file:
<img width="1230" alt="image" src="https://github.com/Baroshem/nuxt-security/assets/1479319/165215c9-6b75-4349-ab5f-2e5b9a5e930b">

Sure enough, if I go into Cloudflare, I can see that it has auto-minify turned on for CSS & JS. I'd bet that's the culprit - thanks guys!

**Comment by Baroshem** [comment]
@georgesaliba Thanks for verifying that.

Would you be able to disable this css/js minify for a while to see if it makes the Nuxt Security 1.0.0-rc.4 work?

This is currently our main blocker for releasing 1.0.0 version :(

**Comment by georgesaliba** [comment]
I disabled the minification at Cloudflare and flushed my cached, but I'm still seeing minified files. Not yet sure if it's just caching (files could be cached in some of the data centers even after flushing) or if there's something else in the stack that might be minifying the file. But I need to get the files matching first or redeploying with SRI enabled would guarantee an outage.

**Comment by Baroshem** [comment]
I see. 

Let us know then if you will make it work.

@vejja Do you think we could publish 1.0.0 tomorrow or should we wait until I will be back ~1st of December?

**Comment by georgesaliba** [comment]
So the cache appears to have flushed, and there are far fewer modifications to the files in production vs. pre-deployment. However, they're still different. It appears something is changing the space character. I'm not sure if it's part of some post-build deployment step (not intentional on my part if so), or some optimization by Google or Cloudflare, or just a result of different systems. But the space characters in the following are different, even though they look the same:

<img width="937" alt="image" src="https://github.com/Baroshem/nuxt-security/assets/1479319/6adbf57b-c927-4eef-8e7c-0ec435cd16cd">

**Comment by vejja** [comment]
Yes, this might be related to 2 different systems used for building.
Is there a way you can build in the production environment but **without deploying**? We would then be able to check manually the hashes without risking your site to go down ?

**Comment by georgesaliba** [comment]
Nope, don't think I have a way to do that.

**Comment by vejja** [comment]
Ok
How do you deploy ? Is there a CI/CD pipeline ?

**Comment by georgesaliba** [comment]
@vejja No, I literally just run `npm run build && firebase deploy` and Firebase takes care of deploying it.

**Comment by vejja** [comment]
> @vejja No, I literally just run `npm run build && firebase deploy` and Firebase takes care of deploying it.

Ok I get it. But then there should be no differences between your local build and the Firebase deployed files, so that's the worse-case scenario from my perspective...

Is there anything specific about your `npm run build` or your `nuxt.config.ts` ? Do you use a Nitro preset to deploy to Firebase, or do you use some kind of `firebaserc` config to indicate where your functions are ?
I would have assumed that `npm run build` would have built an SSR application (unless you modified the package.json build script definition), and I think you said it was SSG/SPA.

I am asking because 
- If it is indeed SSG/SPA, there is something wrong but we can examine your local files in the temporary `.nuxt` folder
- If it is in fact SSR, I would be ready to bet that everything will be fine with integrity hashes.

In other words, I'm trying to have a detailed picture of _who_ is serving your files.

**Comment by vejja** [comment]
One more thing
If I understand correctly, you are deploying your static assets to Firebase Hosting.
Would you be able to log into the Firebase dashboard and inspect the static files to see whether they have whitespace or non-breaking whitespace ?

**Comment by vejja** [comment]
PR #304 adds a section in Documentation about Post-Build Processing.
